chapters/VK_EXT_present_timing/queries.txt

```diff
@@ -0,0 +1,441 @@
+--
+
+[open,refpage='VkPastPresentationTimingEXT',desc='Structure containing timing information about a previously-presented image',type='structs']
+--
+
+The sname:VkPastPresentationTimingEXT structure is defined as:
+
+include::{generated}/api/structs/VkPastPresentationTimingEXT.txt[]
+
+  * pname:sType is the type of this structure.
+  * pname:pNext is `NULL` or a pointer to an extension-specific structure.
+  * pname:presentID is an application-provided value that was given to a
+    previous fname:vkQueuePresentKHR command via
+    slink:VkPresentTimeEXT::pname:presentID.
+    It can: be used to uniquely identify a previous present with the
+    flink:vkQueuePresentKHR command.
+  * pname:targetPresentTime is an application-provided value that was given
+    to a previous flink:vkQueuePresentKHR command via
+    slink:VkPresentTimeEXT::pname:targetPresentTime.
+    If non-zero, it was used by the application to indicate that an image
+    not be presented any sooner than pname:targetPresentTime.
+  * pname:actualPresentTime is the time when the image of the
+    pname:swapchain was actually displayed.
+  * pname:optimalPresentTime is the time when the presentation engine (PE)
+    would have liked the application to have set pname:targetPresentTime to.
+    This allows the PE to provide feedback to the application-provided
+    slink:VkPresentTimeEXT::pname:idealPresentTime.
+    The PE may: set this to pname:actualPresentTime, to
+    slink:VkPresentTimeEXT::pname:idealPresentTime, or to some other
+    time based upon how the application is performing, the system load
+    and/or future clock settings, etc.
+  * pname:timingPropertiesChanged is ename:VK_TRUE if the swapchain's
+    timing properties have changed since the last time those
+    properties were queried with
+    flink:vkGetSwapchainTimingPropertiesEXT, otherwise the properties
+    have not changed.
+    If ename:VK_TRUE, an application must: not compare the values of
+    pname:actualPresentTime and pname:optimalPresentTime with any
+    other values, as the pname:presentation engine may not be able to
+    provide accurate values.
+  * pname:timeDomainChanged is ename:VK_TRUE if the time domain enabled for
+    the swapchain is not currently available.
+    The application must: query what time domains are available and
+    enable display timing with a currently-available time domain.
+    If the currently-enabled time domain is the opaque domain of
+    ename:VK_TIME_DOMAIN_SWAPCHAIN_LOCAL_EXT, it is possible that
+    ename:VK_TIME_DOMAIN_SWAPCHAIN_LOCAL_EXT will be returned by
+    flink:vkGetSwapchainTimeDomainsEXT.
+    In such a case, the presentation engine may: have multiple opaque
+    time domains that it is switching between.
+    If ename:VK_TRUE, an application must: not compare the values of
+    pname:actualPresentTime and pname:optimalPresentTime with any
+    other values, as the pname:presentation engine may not be able to
+    provide accurate values.
+
```

Comment:
Another feature request for this structure: I'd like to have an indicator of how the present was performed and how trustworthy/reliable the reported actualPresentTime is. Given how absolutely critical proper timing and timestamping is for research applications like the ones i mentioned, it is important to be able to assess how likely it is that presentation happened in a reliable way and that the timestamps reflect reality. A research application may want to warn the user or even abort operation if it detects unreliable operating conditions. Or it wants to guide the user in troubleshooting common problems.

E.g., in my experience, on current desktop operating systems like Linux, Windows, macOS, a minimum requirement for being reasonably certain that reported/computed presentation timestamps are of sufficient accuracy and that an image was presented in a pixel exact fashion without any artifacts is that a window is fullscreen exclusive, ie. unobscured, top-level, decorationless, with the desktop compositor mostly out of the way and using gpu page-flipping for presenting images. A research application wants to be certain that its window was not obscured by other windows (or popup notification messages), the pixels were displayed as specified (ie. not altered due to alpha-blending, soft-shadows from other windows, or modified by some desktop effects), without tearing and at the time represented by the timestamps. Anything else would be considered a critical failure in this scenario - something that needs to be reported to the user and fixed by the user immediately.

On Linux/X11/GLX on the XOrg X-Server with FOSS graphics stack, i can use the GLX extension INTEL_swap_events to find out if pageflipping was used for presentation (good!) or if framebuffer copies were used (bad!). With a properietary graphics driver like NVidia's i can still use more painful and non-portable hacks like evaluating return values from GLX_ext_buffer_age and low-level hacks like reading certain gpu MMIO hardware registers directly to try to figure out if pageflipping was used.

On Linux+Wayland, the presentation_time protocol extension provides feedback about how a present happened, at least on the few Wayland compositors that implement the extension. We have flags like
WP_PRESENTATION_FEEDBACK_KIND_HW_CLOCK, WP_PRESENTATION_FEEDBACK_KIND_HW_COMPLETION,
WP_PRESENTATION_FEEDBACK_KIND_ZERO_COPY and
WP_PRESENTATION_FEEDBACK_KIND_VSYNC
to tell us if the present was synchronized to the refresh (KIND_VSYNC) and thereby tear-free, done with a pageflip or equivalent into some hardware overlay plane (KIND_ZERO_COPY), if present completion was detected and signalled by some robust and trustworthy gpu hardware mechanism like some pageflip completion interrupt or similar (KIND_HW_COMPLETION), and timestamped with a high-precision clock (e.g., a gpu hardware timestamp or some mechanism of equivalent reliability and precision == KIND_HW_CLOCK). The Wayland protocol extension is an improvement over the X11 INTEL_swap_events extension, and both are huge improvements over what other operating systems and windowing systems have to offer.

For serious research applications and similar "production" use cases that go beyond pure entertainment like games or video players, it would be great if the VkPastPresentationTimingEXT feedback could improve on these mechanisms, or at least retain some similar information to fulfill the same purpose of reliability assessment or aid in troubleshooting.
Even if some vendor implementations would just return some VK_DONT_KNOW flag to bail on such reporting if the burden of implementation is deemed to high. I would certainly try to contribute at least to Linux FOSS driver implementations of the extension to make sure that such feedback flags would be implemented in a useful way.

Reply:
I have seen a few requests for querying the nitty gritty details of a present (vblank duration, OS present path, hardware accuracy, etc). While I agree that it's useful, I would prefer if this was the subject of a dedicated extension (or a KHR revision of this) which could extend this one, e.g. by adding hardware/platform-specific details to the pNext chain of `VkPastPresentationTimingEXT`. So unless there is a strong push for this, I would prefer to leave these details out for now.